serve/opentelemetry.go

```diff
@@ -146,20 +189,116 @@ func setupOtel(ctx context.Context, logger zerolog.Logger, p *plugin.Plugin, ote
 		metric.WithReader(metricReader),
 		metric.WithResource(pluginResource),
 	)
+
+	lp := log.NewLoggerProvider(
+		log.WithProcessor(logsProcessor),
+		log.WithResource(newResource(p)),
+	)
+
 	otel.SetErrorHandler(otel.ErrorHandlerFunc(func(err error) {
 		logger.Warn().Err(err).Msg("otel error")
 	}))
 	otel.SetTracerProvider(tp)
 	otel.SetMeterProvider(mt)
+	logglobal.SetLoggerProvider(lp)
 
 	shutdown = func() {
 		if err := tp.Shutdown(context.Background()); err != nil {
-			logger.Error().Err(err).Msg("failed to shutdown OTLP trace exporter")
+			logger.Error().Err(err).Msg("failed to shutdown OTLP trace provider")
 		}
 		if err := mt.Shutdown(context.Background()); err != nil {
-			logger.Error().Err(err).Msg("failed to shutdown OTLP metric exporter")
+			logger.Error().Err(err).Msg("failed to shutdown OTLP metric provider")
+		}
+		if err := lp.Shutdown(context.Background()); err != nil {
+			logger.Error().Err(err).Msg("failed to shutdown OTLP logger provider")
 		}
 	}
 
 	return shutdown, nil
 }
+
+// Similar to https://github.com/AkhigbeEromo/opentelemetry-go-contrib/blob/dedcf91a55a36a5a8589c56f2e43c188eb42f4f2/bridges/otelzerolog/hook.go
+// but with `TraceLevel` and attributes support
+type otelLoggerHook struct {
+	otellog.Logger
+	ctx context.Context
+}
+
+func (h *otelLoggerHook) Run(e *zerolog.Event, level zerolog.Level, message string) {
+	record := otellog.Record{}
+	record.SetTimestamp(time.Now().UTC())
+	record.SetSeverity(otellogSeverity(level))
+	record.SetBody(otellog.StringValue(message))
+	// See https://github.com/rs/zerolog/issues/493, this is ugly but it works
+	logData := make(map[string]any)
```

Comment:
Opened https://github.com/rs/zerolog/pull/682 let's see how that goes 😄

Reply:
😱 https://github.com/rs/zerolog/issues/493#issuecomment-1458241050 🤣

Reply:
Can you just add a general comment (not about the hack, but about what's going on with zerolog + otel? It's very clear with the context of this PR but the code looks complex so it's not clear on its own I think.